File: src/utils/ccswitchMatch.test.ts
import { describe, it, expect, beforeAll, afterAll } from "vitest";
import { mockIPC, clearMocks } from "@tauri-apps/api/mocks";
import {
  matchCcProvider,
  extractModels,
  ccProviderToPlatformJson,
  DEFAULT_DIMS,
} from "./ccswitchMatch";
import { __resetDefaultsCacheForTests } from "../domains/platforms";
import type { CcProvider } from "../services/api";

// defaults.json 走 Tauri command 异步读，测试环境 mock 最小数据（deepseek + anthropic + openai）。
// PROTOCOLS async 化后 keywords/name 也从 JSON 派生（buildProtocolsFromPresets），mock 须含字段。
// keywords 对齐 platform-presets.json 真值（anthropic: ["claude","克劳德","官方"]，禁用 "anthropic"
// 避免误匹配 base_url 子串 https://api.deepseek.com/anthropic）。
const DEFAULTS_MOCK = JSON.stringify({
  version: "1",
  last_updated: 0,
  protocols: {
    anthropic: { client_type: "claude_code", keywords: ["claude", "克劳德", "官方"], name: { "en-US": "Anthropic" }, endpoints: { default: [
      { protocol: "anthropic", base_url: "https://api.anthropic.com", client_type: "claude_code" },
    ] }, models: { default: {} }, model_list: { default: [] } },
    openai: { client_type: "codex_tui", keywords: ["gpt", "chatgpt", "官方"], name: { "en-US": "OpenAI" }, endpoints: { default: [
      { protocol: "openai", base_url: "https://api.openai.com/v1", client_type: "codex_tui" },
    ] }, models: { default: {} }, model_list: { default: [] } },
    deepseek: { client_type: "default", keywords: ["深度求索", "deepseek"], name: { "en-US": "DeepSeek" }, endpoints: { default: [
      { protocol: "openai", base_url: "https://api.deepseek.com/v1", client_type: "codex_tui" },
      { protocol: "anthropic", base_url: "https://api.deepseek.com/anthropic", client_type: "claude_code" },
    ] }, models: { default: {} }, model_list: { default: [] } },
  },
});

beforeAll(async () => {
  __resetDefaultsCacheForTests();
  mockIPC((cmd: string) => cmd === "get_defaults_json" ? DEFAULTS_MOCK : null);
  // 见 sub2apiMatch.test.ts：setup.ts afterEach clearMocks 在每测试后清 invoke，
  // docPromise 必须在 mockIPC 仍挂载时缓存好（预热触发 loadDoc 一次）。
  await matchCcProvider({ id: "_warmup", appType: "claude", name: "zzz-unmatched", settingsConfig: {} });
});
afterAll(() => clearMocks());

function claudeProvider(over: Partial<CcProvider> = {}): CcProvider {
  return {
    id: "p1",
    appType: "claude",
    name: "Anthropic",
    settingsConfig: {},
    ...over,
  };
}

function codexProvider(over: Partial<CcProvider> = {}): CcProvider {
  return {
    id: "p2",
    appType: "codex",
    name: "Codex",
    settingsConfig: {},
    ...over,
  };
}

describe("matchCcProvider", () => {
  it("matches via preset keyword (name)", async () => {
    const r = await matchCcProvider(claudeProvider({ name: "deepseek", detectedBaseUrl: "" }));
    expect(r.matchedBy).toBe("preset_keyword");
    expect(r.protocol).toBe("deepseek");
  });

  it("matches via base_url host when no keyword", async () => {
    const r = await matchCcProvider(
      claudeProvider({ name: "Unknown", detectedBaseUrl: "https://api.deepseek.com/anthropic" }),
    );
    expect(["preset_keyword", "base_url_host"]).toContain(r.matchedBy);
    expect(r.protocol).toBe("deepseek");
  });

  it("overrides matched endpoint base_url with provider base_url", async () => {
    const r = await matchCcProvider(
      claudeProvider({ name: "deepseek", detectedBaseUrl: "https://my.proxy/anthropic" }),
    );
    const ep = r.endpoints.find((e) => e.protocol === r.protocol);
    // deepseek protocol type's same-proto endpoint gets overridden
    expect(r.endpoints.length).toBeGreaterThan(0);
    expect(ep === undefined || ep.base_url === "https://my.proxy/anthropic").toBe(true);
  });

  it("codex fallback uses openai_responses when wireApi=responses", async () => {
    const r = await matchCcProvider(
      codexProvider({
        name: "zzz-unmatched",
        detectedBaseUrl: "https://unknown.example.com",
        codexConfigParsed: { wireApi: "responses" },
      }),
    );
    expect(r.matchedBy).toBe("protocol_fallback");
    expect(r.protocol).toBe("openai");
    expect(r.endpoints[0].protocol).toBe("openai_responses");
    expect(r.endpoints[0].client_type).toBe("codex_tui");
  });

  it("codex fallback uses openai when wireApi is not responses", async () => {
    const r = await matchCcProvider(
      codexProvider({ name: "zzz-unmatched", detectedBaseUrl: "https://unknown.example.com" }),
    );
    expect(r.endpoints[0].protocol).toBe("openai");
  });

  it("claude fallback to anthropic when unmatched", async () => {
    const r = await matchCcProvider(
      claudeProvider({ name: "zzz-unmatched", detectedBaseUrl: "https://unknown.example.com" }),
    );
    expect(r.matchedBy).toBe("protocol_fallback");
    expect(r.protocol).toBe("anthropic");
    expect(r.endpoints[0].client_type).toBe("claude_code");
  });

  it("claude fallback to anthropic even when base_url looks openai (newapi)", async () => {
    const r = await matchCcProvider(
      claudeProvider({ name: "zzz-unmatched", detectedBaseUrl: "https://relay.zzqq-unmatched.dev/v1" }),
    );
    expect(r.protocol).toBe("anthropic");
  });

  it("handles empty base_url in fallback", async () => {
    const r = await matchCcProvider(claudeProvider({ name: "zzz-unmatched" }));
    expect(r.protocol).toBe("anthropic");
  });
});

describe("extractModels", () => {
  it("extracts claude env model slots", () => {
    const p = claudeProvider({
      settingsConfig: {
        env: {
          ANTHROPIC_MODEL: "claude-sonnet",
          ANTHROPIC_DEFAULT_HAIKU_MODEL: "claude-haiku",
          ANTHROPIC_DEFAULT_SONNET_MODEL: "claude-sonnet-4",
          ANTHROPIC_DEFAULT_OPUS_MODEL: "claude-opus",
        },
      },
    });
    expect(extractModels(p)).toEqual({
      default: "claude-sonnet",
      haiku: "claude-haiku",
      sonnet: "claude-sonnet-4",
      opus: "claude-opus",
    });
  });

  it("returns empty object when claude env missing", () => {
    expect(extractModels(claudeProvider())).toEqual({
      default: undefined,
      haiku: undefined,
      sonnet: undefined,
      opus: undefined,
    });
  });

  it("extracts codex model", () => {
    const p = codexProvider({ codexConfigParsed: { model: "gpt-5" } });
    expect(extractModels(p)).toEqual({ default: "gpt-5" });
  });

  it("codex without parsed config → empty default", () => {
    expect(extractModels(codexProvider())).toEqual({ default: undefined });
  });
});

describe("ccProviderToPlatformJson", () => {
  it("assembles platform json honouring dims", async () => {
    const provider = claudeProvider({
      name: "deepseek",
      detectedBaseUrl: "https://api.deepseek.com/anthropic",
      detectedApiKey: "sk-key",
      settingsConfig: { env: { ANTHROPIC_MODEL: "m1" } },
    });
    const match = await matchCcProvider(provider);
    const json = ccProviderToPlatformJson(provider, match, DEFAULT_DIMS);
    expect(json.name).toBe("deepseek");
    expect(json.api_key).toBe("sk-key");
    expect(json.base_url).toBe("https://api.deepseek.com/anthropic");
    expect((json.models as { default?: string }).default).toBe("m1");
  });

  it("omits models when d2 off and key when d4 off", async () => {
    const provider = claudeProvider({
      name: "deepseek",
      detectedApiKey: "sk-key",
      settingsConfig: { env: { ANTHROPIC_MODEL: "m1" } },
    });
    const match = await matchCcProvider(provider);
    const json = ccProviderToPlatformJson(provider, match, { d1: true, d2: false, d4: false });
    expect(json.models).toEqual({});
    expect(json.api_key).toBe("");
  });
});

File: flutter/test/settings/ccswitch_match_test.dart
/// cc-switch provider → Platform JSON 的匹配回退链与转换
/// （`lib/src/pages/settings/ccswitch_match.dart`，`src/utils/ccswitchMatch.ts` 的 Dart 版）。
///
/// 回归的是这条：Flutter 侧原先把 cc-switch 的**原始 provider map** 直接当
/// `platformPayload` 发给后端（`importexport_page.dart:718` 的 `(chosen) => chosen`），
/// 而后端收的是已经转好的 Platform JSON，缺字段一律写空串 ——
/// 导进去的是一串没有协议、没有 base_url、没有密钥的空壳平台。
library;

import 'package:aidog_flutter/src/pages/platform_card_bits.dart'
    show ProtocolMetaTable;
import 'package:aidog_flutter/src/pages/platform_defaults.dart';
import 'package:aidog_flutter/src/pages/settings/ccswitch_match.dart';
import 'package:flutter_test/flutter_test.dart';

/// 两份元数据都来自同一份 registry 文档 —— 与页面上的取法一致。
/// glm 带关键词，anthropic 只靠 host（故意不给关键词，用来验步骤 2）。
const String _defaultsJson = '''
{"protocols":{
  "glm":{"name":{"en-US":"Zhipu GLM"},"keywords":["glm","zhipu"],
    "endpoints":{"default":[{"protocol":"glm","base_url":"https://open.bigmodel.cn/api/paas/v4"}]}},
  "anthropic":{"name":{"en-US":"Anthropic"},
    "endpoints":{"default":[{"protocol":"anthropic","base_url":"https://api.anthropic.com"}]}},
  "openai":{"name":{"en-US":"OpenAI"},
    "endpoints":{"default":[{"protocol":"openai","base_url":"https://api.openai.com/v1"}]}}
}}''';

Map<String, Object?> _claude({
  String name = 'x',
  String? baseUrl,
  String? apiKey,
  Map<String, Object?>? env,
}) => {
  'id': 'p1',
  'appType': 'claude',
  'name': name,
  'settingsConfig': {'env': env ?? const <String, Object?>{}},
  'detectedBaseUrl': baseUrl,
  'detectedApiKey': apiKey,
};

Map<String, Object?> _codex({
  String name = 'x',
  String? baseUrl,
  String? wireApi,
  String? model,
}) => {
  'id': 'p2',
  'appType': 'codex',
  'name': name,
  'settingsConfig': const <String, Object?>{},
  'detectedBaseUrl': baseUrl,
  'codexConfigParsed': {'wireApi': wireApi, 'model': model},
};

void main() {
  final meta = ProtocolMetaTable.parse(_defaultsJson, 'en-US');
  final defaults = PlatformDefaults.parse(_defaultsJson, '', 'en-US');

  group('匹配回退链（ccswitchMatch.ts:163）', () {
    test('步骤 1：名字命中 preset 关键词', () {
      final m = matchCcProvider(
        _claude(name: 'my GLM account'),
        meta: meta,
        defaults: defaults,
      );
      expect(m.protocol, 'glm');
      expect(m.matchedBy, CcMatchedBy.presetKeyword);
      expect(m.matchedLabel, 'Zhipu GLM');
    });

    test('步骤 1：名字没线索但 base_url 里有关键词也算命中（name + url 一起当 hay）', () {
      final m = matchCcProvider(
        _claude(name: '账号一', baseUrl: 'https://zhipu.example.com/v1'),
        meta: meta,
        defaults: defaults,
      );
      expect(m.protocol, 'glm');
      expect(m.matchedBy, CcMatchedBy.presetKeyword);
    });

    test('步骤 2：关键词没命中 → 按 base_url 的 host 匹配', () {
      final m = matchCcProvider(
        _claude(name: '账号二', baseUrl: 'https://api.anthropic.com'),
        meta: meta,
        defaults: defaults,
      );
      expect(m.protocol, 'anthropic');
      expect(m.matchedBy, CcMatchedBy.baseUrlHost);
    });

    test('命中之后：preset 骨架端点里同协议那条换成实际 base_url', () {
      const url = 'https://relay.example.com/api/paas/v4';
      final m = matchCcProvider(
        _claude(name: 'glm 中转', baseUrl: url),
        meta: meta,
        defaults: defaults,
      );
      expect(m.endpoints.single.protocol, 'glm');
      expect(m.endpoints.single.baseUrl, url);
      expect(m.baseUrl, url);
    });

    test('步骤 3：codex + wireApi=responses → openai_responses 端点，客户端按 codex 模拟', () {
      final m = matchCcProvider(
        _codex(name: '某中转', baseUrl: 'https://nowhere.test/v1', wireApi: 'responses'),
        meta: meta,
        defaults: defaults,
      );
      expect(m.protocol, 'openai');
      expect(m.matchedBy, CcMatchedBy.protocolFallback);
      expect(m.endpoints.single.protocol, 'openai_responses');
      expect(m.endpoints.single.clientType, 'codex_tui');
    });

    test('步骤 3：codex 没写 wireApi → 退到 openai 端点', () {
      final m = matchCcProvider(
        _codex(name: '某中转', baseUrl: 'https://nowhere.test/v1'),
        meta: meta,
        defaults: defaults,
      );
      expect(m.endpoints.single.protocol, 'openai');
    });

    test('步骤 4：claude 一路没命中 → anthropic，即使 URL 看着像 openai', () {
      final m = matchCcProvider(
        _claude(name: '某中转', baseUrl: 'https://nowhere.test/v1/chat/completions'),
        meta: meta,
        defaults: defaults,
      );
      expect(m.protocol, 'anthropic');
      expect(m.matchedBy, CcMatchedBy.protocolFallback);
      expect(m.endpoints.single.clientType, 'claude_code');
    });

    test('registry 没到手（空表）也不崩，退到协议回退那一支', () {
      final m = matchCcProvider(
        _claude(name: 'my GLM account', baseUrl: 'https://open.bigmodel.cn'),
        meta: const ProtocolMetaTable(),
        defaults: PlatformDefaults.empty,
      );
      expect(m.protocol, 'anthropic');
      expect(m.matchedBy, CcMatchedBy.protocolFallback);
    });
  });

  group('模型映射抽取（ccswitchMatch.ts:200）', () {
    test('claude：四个 env 各进各的槽位', () {
      final m = extractCcModels(
        _claude(
          env: {
            'ANTHROPIC_MODEL': 'claude-x',
            'ANTHROPIC_DEFAULT_HAIKU_MODEL': 'h',
            'ANTHROPIC_DEFAULT_SONNET_MODEL': 's',
            'ANTHROPIC_DEFAULT_OPUS_MODEL': 'o',
          },
        ),
      );
      expect(m.defaultModel, 'claude-x');
      expect(m.haiku, 'h');
      expect(m.sonnet, 's');
      expect(m.opus, 'o');
    });

    test('空串按「没有」处理，不写成空字符串槽位', () {
      final m = extractCcModels(_claude(env: {'ANTHROPIC_MODEL': ''}));
      expect(m.defaultModel, isNull);
    });

    test('codex：取 config.toml 的 model', () {
      expect(extractCcModels(_codex(model: 'gpt-x')).defaultModel, 'gpt-x');
    });
  });

  group('转成 Platform JSON（ccswitchMatch.ts:236）', () {
    Map<String, Object?> convert(
      Map<String, Object?> p, {
      CcImportDims dims = const CcImportDims(),
    }) => ccProviderToPlatformJson(
      p,
      matchCcProvider(p, meta: meta, defaults: defaults),
      dims,
    );

    test('协议 / base_url / 密钥 / 端点都落进去了（这就是「空壳平台」的反面）', () {
      final json = convert(
        _claude(
          name: 'GLM 主号',
          baseUrl: 'https://open.bigmodel.cn/api/paas/v4',
          apiKey: 'sk-test',
        ),
      );
      expect(json['name'], 'GLM 主号');
      expect(json['platform_type'], 'glm');
      expect(json['base_url'], 'https://open.bigmodel.cn/api/paas/v4');
      expect(json['api_key'], 'sk-test');
      expect((json['endpoints']! as List), isNotEmpty);
      expect(json['enabled'], true);
      expect(json['status'], 'enabled');
    });

    test('关掉密钥那一维 → api_key 是空串，其余照旧', () {
      final json = convert(
        _claude(name: 'GLM', apiKey: 'sk-test'),
        dims: const CcImportDims(d4: false),
      );
      expect(json['api_key'], '');
      expect(json['platform_type'], 'glm');
    });

    test('关掉模型那一维 → models 是空表', () {
      final json = convert(
        _claude(name: 'GLM', env: {'ANTHROPIC_MODEL': 'claude-x'}),
        dims: const CcImportDims(d2: false),
      );
      expect(json['models'], isEmpty);
    });

    test('开着模型那一维 → models 带上 default 槽', () {
      final json = convert(
        _claude(name: 'GLM', env: {'ANTHROPIC_MODEL': 'claude-x'}),
      );
      expect((json['models']! as Map)['default'], 'claude-x');
    });
  });
}
